api/maven-api-spi/src/main/java/org/apache/maven/api/spi/PropertyContributor.java

```diff
@@ -33,9 +34,12 @@
 @Consumer
 public interface PropertyContributor extends SpiService {
     /**
-     * Invoked just before session is created with a mutable map that carries collected user properties so far.
+     * Invoked just before session is created.
      *
-     * @param userProperties The mutable user properties, never {@code null}.
+     * @param systemProperties Immutable map of system properties, never {@code null}.
+     * @param userProperties Immutable map of user properties, never {@code null}.
+     * @param topDirectory The path of top directory, never {@code null}.
      */
-    void contribute(Map<String, String> userProperties);
+    Map<String, String> contribute(
+            Map<String, String> systemProperties, Map<String, String> userProperties, Path topDirectory);
```

Comment:
I'd use a `Map<String, String>` for paths, to store all required paths, similar to https://github.com/apache/maven/blob/6ac914d6f97517d827f51afa7aaa41e5acdbe023/maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java#L620-L626, as we probably also need the `rootDirectory`.

Reply:
I like type safe things, so added "proto session"